src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -643,12 +677,112 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
             }
         }
 
-        private static Expression GetValueFromProperty(Expression sourceExpression, string key)
+        private static Func<object?, HttpContext, Task> HandleRequestBodyAndCompileRequestDelegateForForm(
+            Expression responseWritingMethodCall,
+            FactoryContext factoryContext)
+        {
+            Debug.Assert(factoryContext.FormRequestBodyParameter is not null, "factoryContext.FormRequestBodyParameter is null for a form body.");
+
+            var parameterTypeName = TypeNameHelper.GetTypeDisplayName(factoryContext.FormRequestBodyParameter.ParameterType, fullName: false);
+            var parameterName = factoryContext.FormRequestBodyParameter.Name;
+
+            Debug.Assert(parameterName is not null, "CreateArgument() should throw if parameter.Name is null.");
+
+            if (factoryContext.ParameterBinders.Count > 0)
+            {
+                // We need to generate the code for reading from the body or form before calling into the delegate
+                var continuation = Expression.Lambda<Func<object?, HttpContext, object?, object?[], Task>>(
+                responseWritingMethodCall, TargetExpr, HttpContextExpr, BodyValueExpr, BoundValuesArrayExpr).Compile();
+
+                // Looping over arrays is faster
+                var binders = factoryContext.ParameterBinders.ToArray();
+                var count = binders.Length;
+
+                return async (target, httpContext) =>
+                {
+                    // Run these first so that they can potentially read and rewind the body
+                    var boundValues = new object?[count];
+
+                    for (var i = 0; i < count; i++)
+                    {
+                        boundValues[i] = await binders[i](httpContext);
+                    }
+
+                    object? formValue = null;
+                    var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
+                    if (feature?.CanHaveBody == true)
+                    {
+                        if (!httpContext.Request.HasJsonContentType())
```

Comment:
```suggestion
                        if (!httpContext.Request.HasFormContentType)
```

Can we add a test case for minimal APIs accepting both form parameters and `BindAsync` parameters?

There's also a lot of duplication between the `if (factoryContext.ParameterBinders.Count > 0)` branch and the `else` branch. Is there a way we can reduce this duplication both here and in HandleRequestBodyAndCompileRequestDelegateForJson? And/or possibly reduce duplication between HandleRequestBodyAndCompileRequestDelegateForJson and HandleRequestBodyAndCompileRequestDelegateForForm?

Reply:
Sure - I erred away from refactoring away the duplication to start with as the way the Lambdas were structured I assumed there was a perf reason it was the way it was.

It was less duplicative initially as there was only a need for an `else`, but some changes between when I opened this and now meant the 2 branches become 4 on a previous rebase.

Unless there's any specific characteristics of why this bit is the way it is, I'll try and factor it out to reduce the duplication.